clippy_lints/src/casts/cast_lossless.rs

```diff
@@ -21,7 +20,12 @@ pub(super) fn check(
     cast_to_hir: &rustc_hir::Ty<'_>,
     msrv: Msrv,
 ) {
-    if !should_lint(cx, cast_from, cast_to, msrv) {
+    if !should_lint(cx, cast_from, cast_to, msrv)
+        // Checking the expression, cast_from_expr and cast_to_hir share the same context.
+        || [cast_from_expr.span, cast_to_hir.span]
+            .iter()
+            .any(|&span| !expr.span.eq_ctxt(span))
```

Comment:
I wonder if we can make an exception here to still lint if `!expr.span.from_expansion()`? That is, still lint
```rs
foo!() as u32
```
(So basically the new false negative in the tests).

I can see how linting expressions in macro definitions could be a hazard since the macro might be invoked multiple times with different types or expressions which are only sometimes compatible, but I don't think there is a harm if the primary cast expression is in the root context (regardless of what or where `cast_from`/`cast_to` comes from), is there?

Reply:
This is possible for the most part. The `SyntaxContext` for the cast and the target type should be the same. The expression really just needs to be a child context which can be checked with something like `walk_span_to_context(expr.span, cast_ctxt).is_some()`.